distsys/mpcalctx.go

```diff
@@ -20,22 +20,56 @@ var ErrCriticalSectionAborted = errors.New("MPCal critical section aborted")
 // ErrContextClosed will be returned if the context of an archetype is closed.
 var ErrContextClosed = errors.New("MPCal context closed")
 
-// ArchetypeResourceHandle encapsulates a reference to an ArchetypeResource.
-// These handles insulate the end-user from worrying about the specifics of resource lifetimes, logging, and
-// crash recovery scenarios.
-type ArchetypeResourceHandle struct {
-	// a semantic path into the tree of resources stored in MPCalContext
-	// the type TLAValue is used because it is easy to use as an immutable.Map key, and resource indices are TLAValue anyway
-	// in general, this can take two forms:
-	// - "named_resource_name" \* named resources are user-defined and live in a flat namespace
-	// - <<... stack_index, ...>> \* in the case where the resource is local, and possibly part of a recursive procedure call
-	path TLAValue
+var ErrDone = errors.New("A pseudo-error to indicate an archetype has terminated execution normally")
+
+var ErrProcedureFallthrough = errors.New("control has reached the end of a procedure body without reaching a return")
+
+type MPCalJumpTable map[string]MPCalCriticalSection
+
+func MakeMPCalJumpTable(criticalSections... MPCalCriticalSection) MPCalJumpTable {
+	tbl := make(MPCalJumpTable)
+	for _, criticalSection := range criticalSections {
+		tbl[criticalSection.Name] = criticalSection
+	}
+	return tbl
+}
+
+type MPCalCriticalSection struct {
+	Name string
+	Body func(iface ArchetypeInterface) error
 }
 
-func (handle ArchetypeResourceHandle) String() string {
-	return handle.path.String()
+type MPCalProcTable map[string]MPCalProc
+
+func MakeMPCalProcTable(procs... MPCalProc) MPCalProcTable {
+	tbl := make(MPCalProcTable)
+	for _, proc := range procs {
+		tbl[proc.Name] = proc
+	}
+	return tbl
 }
 
+type MPCalProc struct {
+	Name string
+	Label string
+	StateVars []string
+	PreAmble func(iface ArchetypeInterface) error
+}
+
+type MPCalArchetype struct {
+	Name string
+	Label string
+	RequiredRefParams, RequiredValParams []string
+	JumpTable MPCalJumpTable
+	ProcTable MPCalProcTable
+	PreAmble func(iface ArchetypeInterface)
```

Comment:
It's a container for custom variable init code. e.g local var declarations where the value is specified.

It's kind of messy, because it's not really a critical section... see ReadArchetypeResourceLocal